### ISSUE TITLE
Yolov8n and coco dataset-labels inconsistency 

### ISSUE THREAD

**Issue by MLDB541** [ultralytics/ultralytics#14937]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi there. I'd like to train yolov8n on coco dataset 2017. I have downloaded the dataset, train, val, test 2017 and annotations_val train 2017 from https://cocodataset.org/#download. 
Since I got an error regarding Not found labels. I searched for labels and I could download them from https://github.com/ultralytics/yolov5/releases/download/v1.0/coco2017labels.zip. The mentioned error solved. Nevertheless, I have now faced with some issues. 1. The number of labels in train and val folders are less than that of in train2017 and val2017 where images are. 2. Duplication of rows in some labels lead to get warning when I run the code for training yolov8n. 
3. I have not found any solid or reliable repository where there are a complete coco datasets with train, val, 2017 and their labels. It should be noted that I don't use Google colab. I am using my PC to do this task. So, I would be glad if someone, an expert, would kindly advise me in this regard. 
Best wishes. 

### Additional

AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed 
train: Scanning D:\Datasets\COCO\labels\train2017.cache... 117266 images, 1021 backgrounds, 0 corrupt: 100%|██████████| 118287/118287 [00:00<?, ?it/s]train: WARNING  D:\Datasets\COCO\images\train2017\000000099844.jpg: 2 duplicate labels removed
train: WARNING  D:\Datasets\COCO\images\train2017\000000201706.jpg: 1 duplicate labels removed
train: WARNING  D:\Datasets\COCO\images\train2017\000000214087.jpg: 1 duplicate labels removed

**Comment by Y-T-G** [comment]
How are they less than train2017 images?

**Comment by Y-T-G** [comment]
It's the corroborated here:
https://huggingface.co/datasets/rafaelpadilla/coco2017

**Comment by MLDB541** [comment]
> How are they less than train2017 images?

Hi. There are about 118k images for training but about 117k of labels. If you download the labels I provided the link, you would see that.

**Comment by Y-T-G** [comment]
> 117266 images, 1021 backgrounds, 0 corrupt: 100%|

Background images don't have labels.

**Comment by MLDB541** [comment]
> > 117266 images, 1021 backgrounds, 0 corrupt: 100%|
> 
> Background images don't have labels.

Oh! I didn't know that. So I wouldn't have any problems with that number of labels, would I? And how can I get ride of duplication? How can I download the dataset from the page you provided me with? Are there any issues of duplication in the dataset labels of the dataset in your page?

**Comment by MLDB541** [comment]
Indeed the warnings stop processing and I don't have any idea to solve it, unless I remove one by one the duplicated rows from labels.

**Comment by Y-T-G** [comment]
Duplicate labels are removed automatically. You don't have to do anything. All sources of COCO dataset will have the same issue. They are not different.

**Comment by MLDB541** [comment]
Thank you so much for your reply and indeed good news. So, when I run the code it stops at the stage you may see in the above additional info. I provided. What shall I do? Just to wait? And is this omitting automatically duplication is permanently or any time I run the code it happens?

**Comment by MLDB541** [comment]
> Duplicate labels are removed automatically. You don't have to do anything. All sources of COCO dataset will have the same issue. They are not different.

Thank you so much for your reply and indeed good news. So, when I run the code it stops at the stage you may see in the above additional info. I provided. What shall I do? Just to wait? And is this omitting automatically duplication is permanently or any time I run the code it happens?

**Comment by Y-T-G** [comment]
It will cache the labels but wont change the original labels.

If it's getting stuck, it could be due to other issues. Is it using CPU? Or is it idle?

You should post the output of `yolo checks`

**Comment by MLDB541** [comment]
> It will cache the labels but wont change the original labels.
> 
> If it's getting stuck, it could be due to other issues. Is it using CPU? Or is it idle?
> 
> You should post the output of `yolo checks`

Yes, it's getting stuck in that stage, I.e. after removing a few duplications. I have a gpu rtx2080 and the code runs on that. This is a part of Yalm,
# coco.yaml

# The root path to your dataset
path: D:/Datasets/COCO



# Train, validation, and test sets
train: train2017.txt  # Path to the file listing train images
val: val2017.txt      # Path to the file listing validation images
test: test-dev2017.txt  # Path to the file listing test images (optional)

# COCO json annotation files


# Number of classes in the dataset
nc: 80

# Class names in the COCO dataset
names:
  0: person
  1: bicycle
....

I wonder if you could let me know how to get 'yolo checks'!

**Comment by Y-T-G** [comment]
> I wonder if you could let me know how to get 'yolo checks'!

It's a command. Run it in terminal or whichever environment you're using ultralytics from.

**Comment by MLDB541** [comment]
> > I wonder if you could let me know how to get 'yolo checks'!
> 
> It's a command. Run it in terminal or whichever environment you're using ultralytics from.

If I run the code, then it gets stuck, so there wouldn't be a way to run the command yolo checks. So, shall I do it after importing yolo and before running the code?

**Comment by pderrenger** [comment]
Yes, you should run the `yolo checks` command in your terminal before starting your training script. This will help diagnose any potential issues with your setup. If the problem persists, ensure you are using the latest version of the Ultralytics package. If the issue continues, please share the output of the `yolo checks` command for further assistance.

**Comment by MLDB541** [comment]
> Yes, you should run the `yolo checks` command in your terminal before starting your training script. This will help diagnose any potential issues with your setup. If the problem persists, ensure you are using the latest version of the Ultralytics package. If the issue continues, please share the output of the `yolo checks` command for further assistance.

I'm grateful for your advice. I have followed on your suggestion. 
The following is the output of the code where I have considered 'yolo checks'. The code also has been also attached thenafter. 

runfile('D:/Datasets/COCO/yolov8_train2.py', wdir='D:/Datasets/COCO')
C:… \anaconda3\envs\yolov8\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
Ultralytics YOLOv8.2.71  Python-3.9.19 torch-1.11.0 CUDA:0 (NVIDIA GeForce RTX 2080, 8192MiB)
Setup complete  (12 CPUs, 31.9 GB RAM, 166.4/585.4 GB disk)
CUDA is available. Running on GPU.
WARNING  Upgrade to torch>=2.0.0 for deterministic training.
engine\trainer: task=detect, mode=train, model=yolov8n.pt, data=D:/Datasets/COCO/coco.yaml, epochs=10, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=cuda:0, workers=8, project=None, name=yolov8n-coco37, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, bgr=0.0, mosaic=1.0, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=runs\detect\yolov8n-coco37

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]             
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]             
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]               
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]           
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]              
  8                  -1  1    460288  ultralytics.nn.modules.block.C2f             [256, 256, 1, True]           
  9                  -1  1    164608  ultralytics.nn.modules.block.SPPF            [256, 256, 5]                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 12                  -1  1    148224  ultralytics.nn.modules.block.C2f             [384, 128, 1]                 
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 15                  -1  1     37248  ultralytics.nn.modules.block.C2f             [192, 64, 1]                  
 16                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
 17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 18                  -1  1    123648  ultralytics.nn.modules.block.C2f             [192, 128, 1]                 
 19                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 21                  -1  1    493056  ultralytics.nn.modules.block.C2f             [384, 256, 1]                 
 22        [15, 18, 21]  1    897664  ultralytics.nn.modules.head.Detect           [80, [64, 128, 256]]          
Model summary: 225 layers, 3,157,200 parameters, 3,157,184 gradients, 8.9 GFLOPs

Transferred 355/355 items from pretrained weights
Freezing layer 'model.22.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed 
train: Scanning D:\Datasets\COCO\labels\train2017.cache... 117266 images, 1021 backgrounds, 0 corrupt: 100%|██████████| 118287/118287 [00:00<?, ?it/s]train: WARNING  D:\Datasets\COCO\images\train2017\000000099844.jpg: 2 duplicate labels removed
train: WARNING  D:\Datasets\COCO\images\train2017\000000201706.jpg: 1 duplicate labels removed
train: WARNING  D:\Datasets\COCO\images\train2017\000000214087.jpg: 1 duplicate labels removed

**Comment by MLDB541** [comment]
This is the code I have run and obtained above result. 

from ultralytics import YOLO, checks
import torch  # Import torch to check for CUDA and handle device assignment

# Run YOLO environment checks
checks()

# Check if CUDA (GPU) is available
if torch.cuda.is_available():
    print("CUDA is available. Running on GPU.")
else:
    print("CUDA is not available. Running on CPU.")

# Load a YOLOv8n model
model = YOLO('yolov8n.pt')

# Explicitly set the model to run on GPU (if available)
device = 'cuda' if torch.cuda.is_available() else 'cpu'
model.to(device)

# Train the model using the COCO dataset
model.train(
    data='D:/Datasets/COCO/coco.yaml',  # Path to your COCO YAML configuration file
    imgsz=640,
    epochs=10,
    batch=16,
    name='yolov8n-coco'
)

# Validate the model using the validation set
model.val(
    data='D:/Datasets/COCO/coco.yaml'  # Use the same YAML file for validation
)

**Comment by Y-T-G** [comment]
`yolo checks` is a command you run in terminal or in the venv/conda environment you installed in ultralytics.

**Comment by Y-T-G** [comment]
Also use a Python script instead of Jupyter.

**Comment by MLDB541** [comment]
> `yolo checks` is a command you run in terminal or in the venv/conda environment you installed ultralytics in.

This is the output of yolo checks: 

(yolov8) C:\WINDOWS\system32>yolo checks
Ultralytics YOLOv8.2.71 🚀 Python-3.9.19 torch-1.11.0 CUDA:0 (NVIDIA GeForce RTX 2080, 8192MiB)
Setup complete ✅ (12 CPUs, 31.9 GB RAM, 215.8/232.0 GB disk)

OS                  Windows-10-10.0.19045-SP0
Environment         Windows
Python              3.9.19
Install             pip
RAM                 31.94 GB
CPU                 Intel Core(TM) i7-8700K 3.70GHz
CUDA                11.3

numpy               ✅ 1.26.4<2.0.0,>=1.23.0
matplotlib          ✅ 3.9.1>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.13.1>=1.4.1
torch               ✅ 1.11.0>=1.8.0
torchvision         ✅ 0.12.0>=0.9.0
tqdm                ✅ 4.66.4>=4.64.0
psutil              ✅ 5.9.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.0>=2.0.0

(yolov8) C:\WINDOWS\system32>

**Comment by MLDB541** [comment]
> Also use a Python script instead of Jupyter.

I'm using spyder installed on my PC.

**Comment by Y-T-G** [comment]
Launch the training from terminal.

https://docs.ultralytics.com/usage/cli/

**Comment by MLDB541** [comment]
> Launch the training from terminal.
> 
> https://docs.ultralytics.com/usage/cli/

I have tried the approach you mentioned. However, it seems there is an issue in connection. 

Version:1.0StartHTML:00000097EndHTML:00005845StartFragment:00000153EndFragment:00005812<!DOCTYPE><HTML><HEAD></HEAD><BODY><span style="font-size:24px";>(yolov8) C:&gt;yolo detect train data=coco8.yaml model=yolov8n.pt epochs=100 imgsz=640&#13;Downloading https://github.com/ultralytics/assets/releases/download/v8.2.0/yolov8n.pt to 'yolov8n.pt'...&#13;&#9888;&#65039; Download failure, retrying 1/3 https://github.com/ultralytics/assets/releases/download/v8.2.0/yolov8n.pt...&#13;curl: (28) Failed to connect to github.com port 443 after 21054 ms: Couldn't connect to server&#13;Warning: Problem : timeout. 
Will retry in 1 seconds. 3 retries left.&#13;curl: (28) Failed to connect to github.com port 443 after 21055 ms: Couldn't connect to server&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; #&nbsp;&nbsp; # ##&#13;Warning: Problem : timeout.

**Comment by MLDB541** [comment]
Therefore, I've decided to get on with google colab. Nevertheless, it seems Yolo with Coco2017 is not palatable for this environment. As you can see, the images have been all recognized as background. I wonder whether there is a clear way to utilize Yolo+Coco2017 in google colab or a local PC? 
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed ✅

train: Scanning /content/coco/labels/train2017... 0 images, 118287 backgrounds, 0 corrupt: 100%|██████████| 118287/118287 [03:58<00:00, 495.44it/s]

train: WARNING ⚠️ No labels found in /content/coco/labels/train2017.cache. See https://docs.ultralytics.com/datasets for dataset formatting guidance.


train: New cache created: /content/coco/labels/train2017.cache
WARNING ⚠️ No labels found in /content/coco/labels/train2017.cache, training may not work correctly. See https://docs.ultralytics.com/datasets for dataset formatting guidance.
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01), CLAHE(p=0.01, clip_limit=(1, 4.0), tile_grid_size=(8, 8))

/usr/lib/python3.10/multiprocessing/popen_fork.py:66: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  self.pid = os.fork()
val: Scanning /content/coco/labels/val2017... 0 images, 5000 backgrounds, 0 corrupt: 100%|██████████| 5000/5000 [00:10<00:00, 470.39it/s]

val: WARNING ⚠️ No labels found in /content/coco/labels/val2017.cache. See https://docs.ultralytics.com/datasets for dataset formatting guidance.


val: New cache created: /content/coco/labels/val2017.cache
WARNING ⚠️ No labels found in /content/coco/labels/val2017.cache, training may not work correctly. See https://docs.ultralytics.com/datasets for dataset formatting guidance.
Plotting labels to runs/detect/train/labels.jpg...

**Comment by Y-T-G** [comment]
Did you place the labels in the correct folder?

Unless you have Colab Pro, you will not be able to complete training on COCO using Colab.

**Comment by Y-T-G** [comment]
You download the `yolov8n.pt` [manually from GitHub](https://github.com/ultralytics/assets/releases/download/v8.2.0/yolov8n.pt) and place it in the folder where you're starting the training from in your PC.

**Comment by MLDB541** [comment]
> Launch the training from terminal.
> 
> https://docs.ultralytics.com/usage/cli/

Is it possible that we can add some functions to the layers of YOLOV8N through ultralytics tool? If so, how?

**Comment by pderrenger** [comment]
Yes, you can customize YOLOv8 by modifying its layers. You can achieve this by editing the model configuration files or directly modifying the source code in the Ultralytics repository. For detailed guidance, please refer to the Ultralytics documentation on model customization. If you encounter any issues, ensure you are using the latest version of the package.

**Comment by MLDB541** [comment]
> Yes, you can customize YOLOv8 by modifying its layers. You can achieve this by editing the model configuration files or directly modifying the source code in the Ultralytics repository. For detailed guidance, please refer to the Ultralytics documentation on model customization. If you encounter any issues, ensure you are using the latest version of the package.

I'm really grateful for your reply. 
Could you let me know how I can build an exe file from a .py file where some simple commands have been written so that they import YOLO from Ultralytics, use a trained yolov8n.pt and get output from the trained model. It seems PyInstaller is nit able to handle it. Is there a clear example in this regard, or any other wat you may think of. 

Thanks in advance.

**Comment by pderrenger** [comment]
@MLDB541 you're welcome! To create an executable from a Python script that uses YOLOv8, you can use PyInstaller. Ensure you have the latest version of PyInstaller and Ultralytics installed. If PyInstaller isn't working as expected, you might need to adjust the spec file or include additional dependencies manually. Unfortunately, we don't have a specific example for this, but the general approach involves running `pyinstaller your_script.py`. If you encounter issues, please check the PyInstaller documentation for troubleshooting tips.

**Comment by MLDB541** [comment]
> @MLDB541 you're welcome! To create an executable from a Python script that uses YOLOv8, you can use PyInstaller. Ensure you have the latest version of PyInstaller and Ultralytics installed. If PyInstaller isn't working as expected, you might need to adjust the spec file or include additional dependencies manually. Unfortunately, we don't have a specific example for this, but the general approach involves running `pyinstaller your_script.py`. If you encounter issues, please check the PyInstaller documentation for troubleshooting tips.

I cannot thank you enough. I'll take your advice then to see what may happen.

**Comment by pderrenger** [comment]
You're welcome! If you encounter any issues or need further assistance, feel free to ask here. Good luck with your project!
